File: comp-analysis.R
library(dplyr)
library(httr)
library(jsonlite)
library(magrittr)

resp <- GET("https://www.levels.fyi/js/salaryData.json")
cat("Response code: ", resp$status_code, "\n")

if (resp$status_code != 200)
  stop("Failed to fetch data!")

json <- rawToChar(resp$content)
salaries_df <- jsonlite::fromJSON(json) %>%
  filter(
    grepl(", Canada$", location, ignore.case = TRUE),
    title == "Software Engineer",
    as.numeric(yearsofexperience) >= 7,
    as.numeric(yearsofexperience) <= 8
  )

cat("Num samples: ", nrow(salaries_df), "\n")

# (!!) NOTE: all monetary quantities from levels.fyi are stored in USD to
# facilitate comparison of salaries in various countries.
p50 <- median(as.numeric(salaries_df$totalyearlycompensation))
cat("Median: ", p50, "k USD\n")
cat("Standard deviation: ", sd(as.numeric(salaries_df$totalyearlycompensation)), "k USD\n")

# Therefore we need to apply the approximate USD-to-CAD conversion rate as of
# September, 2021.
kUsdToCad <- 1.2647903

# 90-th percentile among SWEs in Canada with 7-8 YOE => ~260k CAD / yr
cat(
  "The percentile: ",
  ecdf(kUsdToCad * as.numeric(salaries_df$totalyearlycompensation))(260 * 1.1 + 6.5) * 100,
  "%\n"
)

File: comp-analysis.py
#!/usr/bin/env python3

from http import HTTPStatus
import pandas as pd
import requests

resp = requests.get("https://www.levels.fyi/js/salaryData.json")
print(f"Response: {resp.status_code}")
if resp.status_code != HTTPStatus.OK:
    # AFAIK this endpoint only responds with HTTP 200 and not any other 2xx
    # status for successful queries
    raise RuntimeError(f"Possible failure in fetching salaryData.json: {resp.status_code}")

data = resp.json()
df = pd.DataFrame(data)
df = df[ \
  df["location"].str.contains(", Canada$") & \
  (df["title"] == "Software Engineer") & \
  (pd.to_numeric(df["yearsofexperience"]) >= 7) & \
  (pd.to_numeric(df["yearsofexperience"]) <= 8) \
]

print(f"Num samples: {df.shape[0]}")
# All monetary quantities are stored in USD in levels.fyi data.
print(f"Median: {pd.DataFrame.median(pd.to_numeric(df.totalyearlycompensation))} k USD")
print(f"Standard deviation: {pd.DataFrame.std(pd.to_numeric(df.totalyearlycompensation))} k USD")

kUsdToCad = 1.28

p90 = pd.to_numeric(df.totalyearlycompensation).quantile(q = 0.9)
print(f"90-th percentile: {p90} k USD")
print(f"90-th percentile: {p90 * kUsdToCad} k CAD")

p75 = pd.to_numeric(df.totalyearlycompensation).quantile(q = 0.75)
print(f"75-th percentile: {p75 * kUsdToCad} k CAD")
